lib/engine/game/g_1848/step/buy_train.rb

```diff
@@ -31,10 +31,10 @@ def can_entity_buy_train?(entity)
           def buyable_trains(entity)
             # Cannot buy 2E if one is already owned
             owns_2e = entity.trains.any? { |t| t.name == '2E' }
-            return super if !owns_2e && @game.phase.status.include?('can_buy_trains')
-
-            super.reject { |t| t.name == '2E' } if owns_2e
-            super.select(&:from_depot?) unless @game.phase.status.include?('can_buy_trains')
+            trains_to_buy = super
```

Comment:
I refactored the code. This is cleaner.
Get all the trains, filter if owns 2e, filter if can't buy from other companies.
Return filtered list if at all.